build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1417,156 @@ message ExecuteRequest {
+  // [StreamedDynamicExecutionData.dynamic_file_hash_response][build.bazel.remote.execution.StreamedDynamicExecutionData].
+  // Depending on the server and client implementation this may be a
+  // relative or fully qualified path.
+  string dynamic_file_hash_path = 5;
+
+  // A file path and expected content hash that the worker does not have in
+  // the CAS. This indicates file content that the `Action` was not predicted
+  // to use from the client, or content that was subsequently removed from
+  // the CAS. The client is requested to upload the file content into the
+  // CAS and respond in 
+  // [StreamedDynamicExecutionData.dynamic_file_content_response][build.bazel.remote.execution.StreamedDynamicExecutionData].
+  DynamicFileContentRequest dynamic_file_content_request = 6;
+}
+
+// A streaming request message for
+// [Execution.ExecuteWithDynamicRetrieval][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval].
+message StreamedDynamicExecutionData {
+  // Sent in response to a dynamic directory metadata request.
+  message DynamicDirectoryMetadataResponse {
+    // The directory path, copied from the directory metadata request.
+    string path = 1;
+
+    // Metadata for the directory. [DirectoryNode][build.bazel.remote.execution.v2.Execution.DirectoryNode]
+    // instances in this field MAY omit digest reference information, as
+    // the directory name is the required information needed to feed into
+    // the filesystem, and the contents of the subdirectory contained in the
+    // `DirectoryNode` are not required to be in the CAS. If the
+    // `DirectoryNode` digest is sent to the worker, the referenced `Directory`
+    // information MUST have been previously uploaded to the CAS.
+    Directory directory_metadata = 2;
+  }
+
+  // Sent in response to a dynamic file hash request.
+  message DynamicFileHashResponse {
+    // The file path, copied from the file hash request.
+    string path = 1;
+    
+    // Set to true if the requested file path does not exist.
+    bool file_does_not_exist = 2;
+
+    // The file digest and associated metadata.
+    FileNode file = 3;
+  }
+
+  // Sent in response to a dynamic file content request.
+  // When `could_not_upload` is false, this message is an acknowledgement
+  // that the client successfully uploaded the requested file content to
+  // the CAS, and the worker should retrieve the file content to resolve
+  // the action's pending file read I/O.
+  message DynamicFileContentResponse {
+    // The file path, copied from the file content request.
+    string path = 1;
+
+    // When true, the file is missing or inaccessible in the client filesystem
+    // and could not be uploaded to the CAS. The worker MAY fail the `Action`
+    // execution with a PRECONDITION_FAILED error, or it MAY return a
+    // file-not-found return code to the executing `Action` file read request
+    // which may cause the action to fail.
+    bool could_not_upload = 2;
```

Comment:
Is there really a need to communicate this back? Couldn't the client just cancel the execution request?

Reply:
There probably should be a way for the client to signal that the client encountered an error reading the file or uploading it to the CAS. I would prefer to only have one option for behaviour here, rather than a choice. I also think I'd prefer it to be an execution failure, since that's what would happen if the client knew that the file was a needed input: it would have to abort upload.